File: Greedy/55_Jump_Game/solution55.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool canJump(vector<int>& nums) {
        int goal = nums.size() - 1;
        
        for (int i = nums.size() - 2; i >= 0; --i) {
            if (i + nums[i] >= goal) {
                goal = i;
            }
        }
        
        return goal == 0;
    }
};

// Runtime1: 36ms, beats 88.62%
// Memory1: 51.15MB, beats 42.11%

// Runtime2: 34ms, beats 93.24%
// Memory2: 51.04MB, beats 80.38%

// Runtime3: 38ms, beats 83.90%
// Memory3: 51.11MB, beats 42.11%

// Runtime reduced by a factor of 10 to not skew data visualization

File: Greedy/55_Jump_Game/solution55.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        goal = len(nums) - 1

        for i in range(len(nums) - 2, -1, -1):
            if i + nums[i] >= goal:
                goal = i
        return goal == 0

# Runtime1: 352ms, beats 65.81%
# Memory1: 17.80MB, beats 68.74%

# Runtime2: 348ms, beats 74.32%
# Memory2: 17.80MB, beats 68.74%

# Runtime3: 343ms, beats 85.72%
# Memory3: 17.79MB, beats 68.74%

# Runtime reduced by a factor of 10 to not skew data visualization